src/utils/price_comparison_utils.ts

```diff
@@ -0,0 +1,110 @@
+import { BigNumber } from '@0x/utils';
+import { Web3Wrapper } from '@0x/web3-wrapper';
+import * as _ from 'lodash';
+
+import { GAS_SCHEDULE_V1 } from '../config';
+import { ZERO } from '../constants';
+import { logger } from '../logger';
+import { ChainId, SourceComparison } from '../types';
+
+import { getTokenMetadataIfExists } from './token_metadata_utils';
+
+const renameIfNativeSource = (source: ERC20BridgeSource) => (source === ERC20BridgeSource.Native ? '0x' : source);
+
+const emptyPlaceholderSources = Object.values(ERC20BridgeSource).reduce<SourceComparison[]>((memo, liquiditySource) => {
+    memo.push({
+        name: renameIfNativeSource(liquiditySource),
+        price: null,
+        gas: null,
+    });
+
+    return memo;
+}, []);
+
+interface PartialRequestParams {
+    buyAmount?: BigNumber;
+    sellAmount?: BigNumber;
+}
+
+interface PartialQuote {
+    buyAmount: BigNumber;
+    sellAmount: BigNumber;
+    buyTokenAddress: string;
+    sellTokenAddress: string;
+    quoteReport?: { sourcesConsidered: QuoteReportSource[] };
+}
+
+export const priceComparisonUtils = {
+    getPriceComparisonFromQuote(
+        chainId: ChainId,
+        params: PartialRequestParams,
+        quote: PartialQuote,
+    ): SourceComparison[] | undefined {
+        if (!quote.quoteReport) {
+            logger.error('Missing quote report, cannot calculate price comparison');
+            return undefined;
+        }
+
+        // NOTE: don't fail quote request if comparison calculations error out
+        try {
+            const buyToken = getTokenMetadataIfExists(quote.buyTokenAddress, chainId);
+            const sellToken = getTokenMetadataIfExists(quote.sellTokenAddress, chainId);
+
+            const isSelling = !!params.sellAmount;
+
+            const { sourcesConsidered } = quote.quoteReport;
+
+            // Filter matching amount samples with a valid result
+            const fullTradeSources = sourcesConsidered.filter(s =>
+                isSelling
+                    ? s.takerAmount.isEqualTo(quote.sellAmount) && s.makerAmount.isGreaterThan(ZERO)
+                    : s.makerAmount.isEqualTo(quote.buyAmount) && s.takerAmount.isGreaterThan(ZERO),
+            );
+
+            // Sort by amount the user will receive and deduplicate to only take the best option for e.g. Kyber
+            const uniqueSources = _.uniqBy(
+                fullTradeSources
+                    .slice()
+                    .sort((a, b) =>
+                        isSelling ? b.makerAmount.comparedTo(a.makerAmount) : a.takerAmount.comparedTo(b.takerAmount),
+                    ),
+                'liquiditySource',
+            );
+
+            const sourcePrices: SourceComparison[] = uniqueSources.map(source => {
+                const { liquiditySource, makerAmount, takerAmount } = source;
+                let gas: BigNumber;
+                if (liquiditySource === ERC20BridgeSource.Native) {
```

Comment:
is there an easy way to ensure this stays up to date when we add new liquidity sources?  or perhaps an existing helper fn we could use?

Reply:
i see -- makes sense -- thanks